phase4-entsog-client/src/main/java/com/helger/phase4/entsog/Phase4ENTSOGSender.java

```diff
@@ -0,0 +1,191 @@
+ * sending, with all potential customization.
+ *
+ * @author Pavel Rotek
+ * @since 0.13.3
+ */
+@Immutable
+public final class Phase4ENTSOGSender {
+	private static final Logger LOGGER = LoggerFactory.getLogger(Phase4ENTSOGSender.class);
+
+	private Phase4ENTSOGSender() {
+	}
+
+	/**
+	 * @return Create a new Builder for AS4 messages if the payload is present.
+	 *         Never <code>null</code>.
+	 */
+	@Nonnull
+	public static ENTSOGUserMessageBuilder builder() {
+		return new ENTSOGUserMessageBuilder();
+	}
+
+	/**
+	 * Abstract ENTSOG UserMessage builder class with sanity methods
+	 *
+	 * @author Philip Helger
+	 * @param <IMPLTYPE> The implementation type
+	 */
+	public abstract static class AbstractENTSOGUserMessageBuilder<IMPLTYPE extends AbstractENTSOGUserMessageBuilder<IMPLTYPE>>
+			extends AbstractAS4UserMessageBuilder<IMPLTYPE> {
+
+		private ECryptoKeyIdentifierType m_signingKeyIdentifierType = ECryptoKeyIdentifierType.ISSUER_SERIAL;
+		private ECryptoKeyIdentifierType m_encryptionKeyIdentifierType = ECryptoKeyIdentifierType.ISSUER_SERIAL;
+		
+		private Phase4OutgoingAttachment m_aPayload;
+		private Charset m_aPayloadCharset;
+		private String m_aPayloadDocumentType;
+		
+		protected AbstractENTSOGUserMessageBuilder() {
+			// Override default values
+			try {
+				httpClientFactory(new Phase4ENTSOGHttpClientSettings());
+			} catch (final Exception ex) {
+				throw new IllegalStateException("Failed to init AS4 Client builder", ex);
+			}
+		}
+
+		/**
+		 * Encryption Key Identifier Type.
+		 *
+		 * @param keyIdentifier ECryptoKeyIdentifierType. Defaults to ISSUER_SERIAL
+		 * @return this for chaining
+		 */
+		public final IMPLTYPE setEncryptionKeyIdentifierType(ECryptoKeyIdentifierType encryptionKeyIdentifierType) {
+			m_encryptionKeyIdentifierType = encryptionKeyIdentifierType;
+			return thisAsT();
+		}
+
+		/**
+		 * Signing Key Identifier Type.
+		 *
+		 * @param keyIdentifier ECryptoKeyIdentifierType. Defaults to ISSUER_SERIAL
+		 * @return this for chaining
+		 */
+		public final IMPLTYPE setSigningKeyIdentifierType(ECryptoKeyIdentifierType signingKeyIdentifierType) {
+			m_signingKeyIdentifierType = signingKeyIdentifierType;
+			return thisAsT();
+		}		
+		
+		/**
+		 * Set the payload to be send out.
+		 *
+		 * @param aBuilder The payload builder to be used.
+		 * @return this for chaining
+		 */
+		@Nonnull
+		public final IMPLTYPE payload(@Nullable final Phase4OutgoingAttachment.Builder aBuilder, Charset aPayloadCharset, String aPayloadDocumentType) {
+			m_aPayload = aBuilder.compressionGZIP().build();
+			m_aPayloadCharset = aPayloadCharset;
+			m_aPayloadDocumentType = aPayloadDocumentType;
+			return thisAsT();
+		}
+
+	    @Override
+	    @OverridingMethodsMustInvokeSuper
+	    public boolean isEveryRequiredFieldSet ()
+	    {
+	      if (!super.isEveryRequiredFieldSet ())
+	        return false;
+
+	      if (m_aPayload==null) {
+	    	  return false;
+	      }
```

Comment:
Is the payloadCharset and the payloadDocumentType mandatory as well? If so a check would be good

Reply:
They are mandatory for EDIGAS types of documents. For other non EDIGAS documents transfered  using ENSOG profile are optional.